### PR TITLE
HPCC-10088 Ganglia Web Service

### PR DESCRIPTION
**Pull request by garonsky**
Signed-off-by: Gleb Aronsky gleb.aronsky@lexisnexis.com

### CODE REVIEW

**Comment by garonsky** [comment]
@miguelvazq
@RussWhitehead please review

**Comment by garonsky** [comment]
@RussWhitehead Changes made based on comments.

**Comment by garonsky** [comment]
@RussWhitehead please review
@GordonSmith please review

File: cmake_modules/FindLIBXML2.cmake [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+#  LIBXML2_INCLUDE_DIR - the libxml2 include directory
+#  LIBXML2_LIBRARIES - The libraries needed to use libxml2
+
+if (NOT LIBXML2_FOUND)
+    SET (libxml2_libs "xml2 libxml2")
+
+  IF (NOT "${EXTERNALS_DIRECTORY}" STREQUAL "")
+    IF (UNIX)
+      IF (${ARCH64BIT} EQUAL 1)
+        SET (osdir "linux64_gcc4.1.1")
```
Comment (GordonSmith) [inline cmake_modules/FindLIBXML2.cmake:31]: Are these meant to be specific versions?

File: CMakeLists.txt [hunk 2]
```diff
@@ -18,8 +18,34 @@ if (NOT LICENSE_FILE)
...
+MESSAGE (HPCC_SOURCE_DIR: ${HPCC_SOURCE_DIR})
+ADD_SUBDIRECTORY(${HPCC_SOURCE_DIR} oss EXCLUDE_FROM_ALL)
+CONFIGURE_FILE(${HPCC_SOURCE_DIR}/build-config.h.cmake "build-config.h" )
+CONFIGURE_FILE(${HPCC_GM_SOURCE_DIR}/scripts/getMetrics.sh.in "getMetrics.sh" )
+INSTALL ( PROGRAMS ${CMAKE_CURRENT_BINARY_DIR}/getMetrics.sh DESTINATION sbin COMPONENT Runtime)
+
+
+HPCC_ADD_SUBDIRECTORY (esp)
+HPCC_ADD_SUBDIRECTORY (initfiles)
+#HPCC_ADD_SUBDIRECTORY (scripts)
```
Comment (GordonSmith) [inline CMakeLists.txt]: Commented out code.

File: cmake_modules/gmCommonSetup.cmake [hunk 3]
```diff
@@ -32,8 +40,128 @@ IF ("${COMMONSETUP_DONE}" STREQUAL "")
 
   cmake_policy ( SET CMP0011 NEW )
 
+  option(CLIENTTOOLS "Enable the building/inclusion of a Client Tools component." OFF)
+  option(PLATFORM "Enable the building/inclusion of a Platform component." ON)
```
Comment (GordonSmith) [inline cmake_modules/gmCommonSetup.cmake]: Why are these here?

**Comment by GordonSmith** [comment]
Looks like a lot of files just copied from Platform - couldn't you just include Platform as a dependency for building?

File: esp/services/ws_rrd/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+         ${HPCC_SOURCE_DIR}/esp/platform
+         ${HPCC_SOURCE_DIR}/system/jlib
+         ${HPCC_SOURCE_DIR}/common/environment
+         ${HPCC_SOURCE_DIR}/esp/services
+         ${HPCC_SOURCE_DIR}/common
+         ${HPCC_SOURCE_DIR}/system/security/securesocket
+         ${HPCC_SOURCE_DIR}/system/security/shared
+         ${HPCC_SOURCE_DIR}/system/include
+         ${HPCC_SOURCE_DIR}/common/remote
+         ${HPCC_SOURCE_DIR}/clients
```
Comment (GordonSmith) [inline esp/services/ws_rrd/CMakeLists.txt:49]: Does this folder exist?

File: esp/services/ws_rrd/RRDGraphWrapper.cpp [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+
+    PROGLOG("RRDTOOL GRAPH CMD --> %s <--", strRRDGraphCmd.str());
+    FILE *fp = popen(strRRDGraphCmd.str(), "r");
+
+    if (fp == NULL)
+    {
+        return NULL;
+    }
+
+    pclose(fp);
```
Comment (GordonSmith) [inline esp/services/ws_rrd/RRDGraphWrapper.cpp:137]: You don't do anything with fp (or are you just checking if the file exists?)

File: esp/services/ws_rrd/RRDGraphWrapper.cpp [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+    pStart = strstr(const_cast<char*>(pBufferIn), "id=\"");
+    pEnd = strstr(pStart+nOffSet, "\"");
+
+    while (pStart != NULL)
+    {
+        int len = pEnd-pStart-nOffSet;
+
+        if (len > 1024)
+        {
+            PROGLOG("SVG id too long");
```
Comment (GordonSmith) [inline esp/services/ws_rrd/RRDGraphWrapper.cpp:201]: Curious as to what the SVG is at this point in the code?  (1024 is not very long)

File: esp/services/ws_rrd/RRDGraphWrapper.cpp [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+            return NULL;
```
Comment (GordonSmith) [inline esp/services/ws_rrd/RRDGraphWrapper.cpp:202]: pBufferIn will leak memory

File: esp/src/ganglia/GangliaFilterDropDownWidget.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+############################################################################## */
+define([
+    "dojo/_base/declare",
+    "dojo/_base/lang",
+    "dojo/i18n",
+    "dojo/i18n!./nls/ganglia",
+    "dojo/_base/array",
+    "dojo/dom",
+    "dojo/dom-form",
+    "dojo/on",
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaFilterDropDownWidget.js]: on not used

File: esp/src/ganglia/GangliaFilterDropDownWidget.js [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        close: function (event) {
+            this.filterDropDown.closeDropDown();
+        },
+        
+        disable: function(disable) {
+            this.filterDropDown.set("disabled", disable);
+        },
+
+        refreshState: function () {
+            //this.iconFilter.src = this.exists() ? dojoConfig.getImageURL("filter.png") : dojoConfig.getImageURL("noFilter.png");
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaFilterDropDownWidget.js]: Commented out code.

File: esp/src/ganglia/GangliaWidget.js [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+define([
+    "dojo/_base/declare",
+    "dojo/_base/array",
+    "dojo/_base/lang",
+    "dojo/i18n",
+    "dojo/i18n!./nls/ganglia",
+    "dojo/dom",
+    "dojo/_base/array",
+    "dojo/dom-construct",
+    "dojo/dom-attr",
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Not used

File: esp/src/ganglia/GangliaWidget.js [hunk 11]
```diff
@@ -0,0 +1,439 @@
...
+    "dojo/promise/all",
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Not Used

File: esp/src/ganglia/GangliaWidget.js [hunk 12]
```diff
@@ -0,0 +1,439 @@
...
+    "dojo/store/Memory",
+    "dojo/request/xhr",
+    "dojo/query",
+
+    "dijit/registry",
+    "dijit/Menu",
+    "dijit/MenuItem",
+
+    "dgrid/OnDemandGrid",
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Is dgrid used directly?

File: esp/src/ganglia/GangliaWidget.js [hunk 13]
```diff
@@ -0,0 +1,439 @@
...
+                context.server = newValue;
+                context._buildMetrics();
+            });
+
+            this.metricsTargetSelect.on('change', function (newValue) {
+                context.metrics = newValue;
+            });
+
+            this.filter.on("clear", function (evt) {
+                context._onFilterType();
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Method does not exist?

File: esp/src/ganglia/GangliaWidget.js [hunk 14]
```diff
@@ -0,0 +1,439 @@
...
+                context.refreshGrid();
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Method does not exist?

File: esp/src/ganglia/GangliaWidget.js [hunk 15]
```diff
@@ -0,0 +1,439 @@
...
+            });
+            this.filter.on("apply", function (evt) {
+                context._onFilterApply();
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js:115]: Method does not exist?

File: esp/src/ganglia/GangliaWidget.js [hunk 16]
```diff
@@ -0,0 +1,439 @@
...
+            this.fromGanliaDateRange.on('change', function (newValue) {
+                context._calculateEpoch(newValue);
+            });
+        },
+
+         initTab: function () {
+            var currSel = this.getSelectedChild();
+        },
+
+        stringTrim: function () {
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Not used/needed?

File: esp/src/ganglia/GangliaWidget.js [hunk 17]
```diff
@@ -0,0 +1,439 @@
...
+
+            this.fromGanliaDateRange.set(name, value);
+
+            this.fromGanliaDateRange.on('change', function (newValue) {
+                context._calculateEpoch(newValue);
+            });
+        },
+
+         initTab: function () {
+            var currSel = this.getSelectedChild();
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Does nothing

File: esp/src/ganglia/GangliaWidget.js [hunk 18]
```diff
@@ -0,0 +1,439 @@
...
+        _buildClusters: function () {
+            var context = this;
+            this.clusterTargetSelect.required = false;
+            WsRrd.GangliaClusterList({
+                request:{
+                    Cluster:"",
+                    Server:""
+                },
+            }).then(function (response) {
+                if (lang.exists("GetAvailableClustersResponse", response)) {
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Should test deeper into response.

File: esp/src/ganglia/GangliaWidget.js [hunk 19]
```diff
@@ -0,0 +1,439 @@
...
+            });
+        },
+
+        _buildServers: function (newValue) {
+            var context = this;
+            console.log(this.serverTargetSelect);
+            this.serverTargetSelect.required = false;
+            WsRrd.GangliaServerList({
+                request:{
+                    Cluster: context.cluster
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Should be this

File: esp/src/ganglia/GangliaWidget.js [hunk 20]
```diff
@@ -0,0 +1,439 @@
...
+                },
+            }).then(function (response) {
+                if (lang.exists("GetAvailableServersForMetricsResponse", response)) {
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: SHould test deeper into response.

File: esp/src/ganglia/GangliaWidget.js [hunk 21]
```diff
@@ -0,0 +1,439 @@
...
+                }
+            });
+        },
+
+        _buildMetrics: function (newValue) {
+            var context = this;
+            this.metricsTargetSelect.required = true;
+            WsRrd.GangliaMetricList({
+                request:{
+                    Cluster:context.cluster,
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: should be this

File: esp/src/ganglia/GangliaWidget.js [hunk 22]
```diff
@@ -0,0 +1,439 @@
...
+                    Server:context.server
+                },
+            }).then(function (response) {
+                if (lang.exists("GetAvailableMetricsResponse.Metrics", response)) {
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: SHould test deeper into response

File: esp/src/ganglia/GangliaWidget.js [hunk 23]
```diff
@@ -0,0 +1,439 @@
...
+
+        _buildTabs: function () {
+            var context = this;
+            var tabs = registry.byId(this.id + "TabContainer");
+            xhr("esp/files/ganglia/ganglia.json", {
+                handleAs: "json"
+            }).then(function (data){
+                arrayUtil.forEach(data.tabs, function(tabItem, idx) {
+                    arrayUtil.forEach(tabItem.metrics, function(metricItem) {
+                        arrayUtil.forEach(tabItem.time, function(timeItem) {
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Is this nesting correct?

File: esp/src/ganglia/GangliaWidget.js [hunk 24]
```diff
@@ -0,0 +1,439 @@
...
+                            var clean = metricItem.join('\n');
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: clean could be calculated in an outer loop rather than every time.

File: esp/src/ganglia/GangliaWidget.js [hunk 25]
```diff
@@ -0,0 +1,439 @@
...
+                            var cleanId = tabItem.name.split(' ').join('_');
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Should be calculated in an outer loop, instead of every time.

File: esp/src/ganglia/GangliaWidget.js [hunk 26]
```diff
@@ -0,0 +1,439 @@
...
+                                        class:'left'
+                                    }, tabItem.name+idx);
+                                }
+                            });
+                        });
+                    });
+                    var bordercontainer = new BorderContainer({
+                        style: "height: 100%; width: 100%;",
+                        title: tabItem.name.split(' ').join('_'),
+                        id: tabItem.name.split(' ').join('_'),
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Could re-use cleanID once moved to an outer loop.

File: esp/src/ganglia/GangliaWidget.js [hunk 27]
```diff
@@ -0,0 +1,439 @@
...
+                    });
+
+                    bordercontainer.addChild(contentpane);
+                    bordercontainer.addChild(toolbar);
+                    toolbar.addChild(button);
+                    bordercontainer.placeAt(context.tabContainer, idx);
+                    bordercontainer.startup();
+                });
+                context.tabContainer.selectChild(data.tabs[0].name.split(' ').join('_'));
+                /*setInterval(function () {
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Commented out code

File: esp/src/ganglia/GangliaWidget.js [hunk 28]
```diff
@@ -0,0 +1,439 @@
...
+
+        _genGraphRefresh : function (id) {
+            var context = this;
+            xhr("esp/files/ganglia/ganglia.json", {
+                handleAs: "json"
+            }).then(function (data){
+                arrayUtil.forEach(data.tabs, function(tabItem, idx) {
+                    arrayUtil.forEach(tabItem.metrics, function(metricItem) {
+                        arrayUtil.forEach(tabItem.time, function(timeItem) {
+                            var clean = metricItem.join('\n');
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js:306]: Could be moved up a loop to prevent re-calculation.

File: esp/src/ganglia/GangliaWidget.js [hunk 29]
```diff
@@ -0,0 +1,439 @@
...
+            });
+        },
+
+        _onRefresh: function (id) {
+
+        },
+
+        _onFilterApply: function () {
+            var context = this;
+            var cluster = context.cluster;
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: should be "this" (for the next few lines)

File: esp/src/ganglia/GangliaWidget.js [hunk 30]
```diff
@@ -0,0 +1,439 @@
...
+            WsRrd.GangliaRRDGraphList({
+                request:{
+                    Clusters: cluster,
+                    Servers: server,
+                    RRDMetrics: metrics,
+                    StartTime: epochFilter,
+                    EndTime: epochNow,
+                    Width: 300,
+                    Height: 120,
+                    Title: context.cluster + ":" + context.server + ":" + context.metrics
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Should be "this"

File: esp/src/ganglia/GangliaWidget.js [hunk 31]
```diff
@@ -0,0 +1,439 @@
...
+                },
+            }).then(function (response) {
+                if (dojo.byId(context.cluster+context.server+context.metrics+context.epochFilter)) {
+                    alert(this.i18n.GraphExists);
+                } else {
+                    if (lang.exists ("GraphSVGDataResponse.Graph", response)) {
+                        var graph = domConstruct.create("div", {
+                            id: context.cluster+context.server+context.metrics+context.epochFilter,
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Consider commenting up id to a separate var

File: esp/src/ganglia/WsRrd.js [hunk 32]
```diff
@@ -0,0 +1,75 @@
...
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+############################################################################## */
+define([
+    "dojo/_base/declare",
+    "dojo/_base/lang",
```
Comment (GordonSmith) [inline esp/src/ganglia/WsRrd.js:18]: Several items declared but not used...

File: esp/src/ganglia/WsRrd.js [hunk 33]
```diff
@@ -0,0 +1,75 @@
...
+    "hpcc/ESPBase",
+    "hpcc/ESPRequest"
+], function (declare, lang, arrayUtil, Deferred, QueryResults, JsonRest, Memory, Cache, Observable,
+    parser,
+    ESPBase, ESPRequest) {
+    var GangliaMetricStore = declare([ESPRequest.Store], {
+        service: "ws_rrd",
+        action: "getAvailableMetrics",
+        responseQualifier: "GetAvailableMetricsResponse.Metrics",
+        idProperty: "Item",
```
Comment (GordonSmith) [inline esp/src/ganglia/WsRrd.js:38]: Is Item an ID or an Object?

File: esp/src/ganglia/WsRrd.js [hunk 34]
```diff
@@ -0,0 +1,75 @@
...
+    });
+
+    var GangliaServerStore = declare([ESPRequest.Store], {
+        service: "ws_rrd",
+        action: "getAvailableServers",
+        responseQualifier: "GetAvailableServersForMetricsResponse.Servers",
+        idProperty: "Item",
```
Comment (GordonSmith) [inline esp/src/ganglia/WsRrd.js:45]: Is Item valid for an ID (or is it an Object?)

File: esp/src/ganglia/WsRrd.js [hunk 35]
```diff
@@ -0,0 +1,75 @@
...
+    });
+
+    return {
+        GangliaClusterList: function (params) {
```
Comment (GordonSmith) [inline esp/src/ganglia/WsRrd.js:49]: As a rule of thumb this file name should equal the Service name and this method name should match the service action.  IOW this file should be "ws_rrd.js" and this function should be "getAvailableClusters"

File: esp/src/ganglia/nls/es/ganglia.js [hunk 36]
```diff
@@ -0,0 +1,13 @@
+define(
+({
+    AddMetric: "",
```
Comment (GordonSmith) [inline esp/src/ganglia/nls/es/ganglia.js]: Its better to exclude any untranslated variables as they will fall back to use the English (otherwise they will display blank)

File: esp/src/ganglia/templates/GangliaWidget.html [hunk 37]
```diff
@@ -0,0 +1,27 @@
...
+                            <option value="Week" id="${id}Week">Week</option>
+                            <option value="Day" id="${id}Day">Day</option>
+                            <option value="Hour" id="${id}Hour">Hour</option>
+                        </select>
+                    </div>
+                    <span data-dojo-type="dijit.ToolbarSeparator"></span>
+                    <!-- <div id="${id}NewPage" class="right" data-dojo-attach-event="onClick:_onNewPage" data-dojo-props="iconClass:'iconNewPage', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.OpenInNewpage}</div> -->
+                </div>
+                <div data-dojo-props="region: 'center'" data-dojo-type="dijit.layout.ContentPane">
+                    <div id="graphs"></div>
```
Comment (GordonSmith) [inline esp/src/ganglia/templates/GangliaWidget.html]: ID Won't be unique if opened in two different places...

File: esp/services/ws_rrd/CMakeLists.txt (hunk 4 above)
Comment (garonsky) [inline esp/services/ws_rrd/CMakeLists.txt:49]: Yes.  This references the platfrom using HPCC_SOURCE_DIR

File: esp/services/ws_rrd/RRDGraphWrapper.cpp (hunk 5 above)
Comment (garonsky) [inline esp/services/ws_rrd/RRDGraphWrapper.cpp:137]: Before closing the stream, through pclose,  I want to ensure that I have a valid FILE pointer.

File: esp/services/ws_rrd/RRDGraphWrapper.cpp (hunk 6 above)
Comment (garonsky) [inline esp/services/ws_rrd/RRDGraphWrapper.cpp:201]: The ID is short.  Looks like a variant of "glyph0-0".  1024 is more than enough for all IDs that I have encountered.

File: cmake_modules/FindLIBXML2.cmake (hunk 1 above)
Comment (garonsky) [inline cmake_modules/FindLIBXML2.cmake:31]: This is the same code as is used by the Platform.

File: cmake_modules/gmCommonSetup.cmake (hunk 3 above)
Comment (garonsky) [inline cmake_modules/gmCommonSetup.cmake]: CLIENTTOOLS and PLATFORM are not needed.  During the build some of these tags are needed by the platform.  Created a Jira to clean up cmake files for 5.4 https://track.hpccsystems.com/browse/HPCC-12892

File: esp/services/ws_rrd/CMakeLists.txt (hunk 4 above)
Comment (garonsky) [inline esp/services/ws_rrd/CMakeLists.txt:49]: The platform is a dependency for building.  This project is built similar to LN, with platform side by side.

**Comment by garonsky** [comment]
@GordonSmith I have accepted all your comments unless I commented otherwise.
@@miguelvazq is working on javascript changes.

File: esp/src/ganglia/GangliaWidget.js (hunk 15 above)
Comment (miguelvazq) [inline esp/src/ganglia/GangliaWidget.js:115]: This method is on line 348

File: esp/src/ganglia/WsRrd.js (hunk 33 above)
Comment (miguelvazq) [inline esp/src/ganglia/WsRrd.js:38]: id

File: esp/src/ganglia/WsRrd.js (hunk 34 above)
Comment (miguelvazq) [inline esp/src/ganglia/WsRrd.js:45]: id

**Comment by miguelvazq** [comment]
@GordonSmith Please re-review. Any other pending issues please assign a JIRA ticket for bug fixes in rc2 or greater.

**Comment by miguelvazq** [comment]
@stuartort FYI

File: esp/src/ganglia/GangliaWidget.js [hunk 38]
```diff
@@ -0,0 +1,422 @@
...
+                context.server = newValue;
+                context._buildMetrics();
+            });
+
+            this.metricsTargetSelect.on('change', function (newValue) {
+                context.metrics = newValue;
+            });
+
+            this.filter.on("clear", function (evt) {
+
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Your filter doesn't emit a clear?

File: esp/src/ganglia/GangliaWidget.js [hunk 39]
```diff
@@ -0,0 +1,422 @@
...
+            });
+
+            this.filter.on("clear", function (evt) {
+                
+            });
+            this.filter.on("apply", function (evt) {
+                context._onFilterApply();
+            });
+
+            this.fromGanliaDateRange.set(name, value);
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: "name" and "value" are undefined?

File: esp/src/ganglia/GangliaWidget.js [hunk 40]
```diff
@@ -0,0 +1,422 @@
...
+                context._calculateEpoch(newValue);
+            });
+        },
+
+        _calculateEpoch: function(newValue){
+            var context = this;
+            this.epochNow = Math.round(new Date().getTime()/1000.0);
+                switch(newValue) {
+                    case "Year":
+                        context.epochFilter = context.epochNow - 31556926
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Missing ";" and no need fot "context" this is in scope...

File: esp/src/ganglia/GangliaWidget.js [hunk 41]
```diff
@@ -0,0 +1,422 @@
...
+                    break;
+                    case "Month":
+                        context.epochFilter = context.epochNow - 2629743
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Missing ";"

File: esp/src/ganglia/GangliaWidget.js [hunk 42]
```diff
@@ -0,0 +1,422 @@
...
+                    break;
+                    case "Week":
+                        context.epochFilter = context.epochNow - 604800
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Missing ";"

File: esp/src/ganglia/GangliaWidget.js [hunk 43]
```diff
@@ -0,0 +1,422 @@
...
+                    break;
+                    case "Day":
+                        context.epochFilter = context.epochNow - 86400
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Missing ";"

File: esp/src/ganglia/GangliaWidget.js [hunk 44]
```diff
@@ -0,0 +1,422 @@
...
+                    break;
+                    case "Hour":
+                        context.epochFilter = context.epochNow - 3600
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Missing ";"

File: esp/src/ganglia/GangliaWidget.js [hunk 45]
```diff
@@ -0,0 +1,422 @@
...
+                        data: output
+                    });
+                    context.clusterTargetSelect.set("store", myStore);
+                }
+            });
+        },
+
+        _buildServers: function (newValue) {
+            var context = this;
+            console.log(this.serverTargetSelect);
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: Should be removed.

File: esp/src/ganglia/GangliaWidget.js [hunk 46]
```diff
@@ -0,0 +1,422 @@
...
+                    });
+                    context.metricsTargetSelect.set("store", myStore);
+                }
+            });
+        },
+
+        _buildTabs: function () {
+            var context = this;
+            var tabs = registry.byId(this.id + "TabContainer");
+            xhr("esp/files/ganglia/ganglia.json", {
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js]: I wonder if that should be "/esp/files/ganglia/ganglia.json" - additional "/" at the start?

File: esp/src/ganglia/GangliaWidget.js [hunk 47]
```diff
@@ -0,0 +1,422 @@
...
+                handleAs: "json"
+            }).then(function (data){
+                arrayUtil.forEach(data.tabs, function(tabItem, idx) {
+                    var cleanId = tabItem.name.split(' ').join('_');
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js:228]: Might be better to use "context.createChildTabID(tabItem.name)"

File: esp/src/ganglia/GangliaWidget.js [hunk 48]
```diff
@@ -0,0 +1,422 @@
...
+
+        },
+
+        _onFilterApply: function () {
+            var context = this;
+            var cluster = this.cluster;
+            var server = this.server;
+            var metrics = this.metrics;
+            var epochFilter = this.epochFilter;
+            var epochNow = this.epochNow;
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js:339]: No real need for:
-            var cluster = this.cluster;
-            var server = this.server;
-            var metrics = this.metrics;
-            var epochFilter = this.epochFilter;
-            var epochNow = this.epochNow;

File: esp/src/ganglia/GangliaWidget.js [hunk 49]
```diff
@@ -0,0 +1,422 @@
...
+            xhr("esp/files/ganglia/ganglia.json", {
+                handleAs: "json"
+            }).then(function (data){
+                WsRrd.GangliaRRDGraphList({
+                request:{
+                    Clusters: cluster,
+                    Servers: server,
+                    RRDMetrics: metrics,
+                    StartTime: epochFilter,
+                    EndTime: epochNow,
```
Comment (GordonSmith) [inline esp/src/ganglia/GangliaWidget.js:397]: Aren't the following all undefined?:
-                    Clusters: cluster,
-                    Servers: server,
-                    RRDMetrics: metrics,
-                    StartTime: epochFilter,
-                    EndTime: epochNow,

File: esp/src/ganglia/GangliaWidget.js (hunk 48 above)
Comment (miguelvazq) [inline esp/src/ganglia/GangliaWidget.js:339]: Good catch for repetitiveness. I also modified the alert message was trying to this.i18n instead of context.i18n and not displaying the string

File: esp/src/ganglia/GangliaWidget.js (hunk 47 above)
Comment (miguelvazq) [inline esp/src/ganglia/GangliaWidget.js:228]: Would have to test further will open a ticket for this improvement.

File: esp/src/ganglia/GangliaWidget.js (hunk 38 above)
Comment (miguelvazq) [inline esp/src/ganglia/GangliaWidget.js]: I removed it altogether. The FilterDropDownWidget does the clear for me.

**Comment by miguelvazq** [comment]
@GordonSmith Please review

**Comment by miguelvazq** [comment]
@GordonSmith modified parameters being undefined.

**Comment by GordonSmith** [comment]
Looks ok now...
